### DISCUSSION TITLE
[RFC Proposal] Stop updating ``Transition``'s slot while a transition is taking place

### DISCUSSION THREAD

**Discussion by ferferga** [vuejs/rfcs#501] [General]
Hello. I'm opening this with the aim to gather some feedback on whether there's interest in having this in Vue's core. In my opinion, my proposal fits into the "small enough" area to not go into the RFC process yet (besides, I don't have a PR yet), and that's why I'm opening it under the "General" topic (but feel free to let me know or to move this discussion there) .

# Motivation/Summary 

When a component is transitioning, state mutations can trigger DOM changes. These can sometimes break the appearance of animations, like in this snippet:

[Vuetify Playground](https://play.vuetifyjs.com/#eNqlUk1v2zAM/SusLk4Af6BbgQFZMmSXYb9gO9Q9ODadapElQ6KTDYH/+yj5o256KdCLLfKR75FPeryK722bnjsUG7ElbFpVEH7LNcD2nBRtG44haAqpxyjEhH8pqSWqChg0FapdLhp3zAVky7oDadiXSpYnxh0xfdqgc8URYQehfq4G+CEtgtNFeToUdibJAsuSdCpZSvs/i29Ko4ydtULEaWXKgqTRDBwMkWlAYU2zeE45Xa/wer6+96CH/AiT5mQJp2ZPfDC4tc0WLnLoSitbAofUeVg2rbEEV7BYx/wpSpJnjKE0TdsRVjFcCiqfoYfa8owR30zkeUqjHXm72DRuXUU/USkDv41V1V20ngrC/KFkIF5d/XTjPhuIotjHwROO0Fpjmb5ff50VvItMMM0zEByRVmsIR2Bu6qy+seput2N2pgHogwYvvNJ4+VWoDudWWcOchB231IVyLzB33byPkfIFCqMvgXBF/Ak7bLPBbjZaxGLwmi+pTf84o/mFB518BFwuNpNyLthoH+fimah1mywrK81t7IY821QjZbptsj2XZbbTJBtMKtPsP6ef0ocvWSUdLfMpuiY5WHNxaJklF8GTUYdk/e8dWmMlS9yn9w+DxJhLVHFwXuMNd8aCZ7SJRV2hRX7579vppu3VXjfYm92mG2DLldQnd+N26YLTj9OUH1s6sHmqJ6/Yx2KYlLWHFvH0H9XemDE=)

*Fire the snackbar and wait a few seconds for it to close. Check how the color of the snackbar changes to white while the transition is taking place. More backrgound here: https://github.com/vuetifyjs/vuetify/issues/17180*

Here's another example where this problem might be annoying. In this example, we will deal with a large list (10000 items) and we want to toggle the visibility using `v-show`, but also kill all the DOM nodes when they're hidden to free up resources:

1) In the first case, we simply have a transition that fires after a ``v-show`` change. In DevTools, you can see how all the nodes are still present in the tree: [Vue Playground](https://play.vuejs.org/#eNqFkk1v2zAMhv8Kp0sTwB9Jt2GD5xYZtuuwSw876OIPJtVifUCi3WWB//soO82yDGhzCCSSet6XNI/is3PZ0KMoRBkarxxBQOrdvTRKO+sJjuBxCyNsvdVww6U30kjTWBMItG2xg7tYsSDf41KaMp8xDOALoXZdRcg3gLLuiayBTdOpZn8nxfPzN9NBiqkK4MHudh3CoIKqVafokMGX3ns0VMDxeBIdxwmZz8yZ/+ArExQp1jCVRlYInWpRCqicw8qf+GWrBhjS8Gifnk2ctTnrOLe1nlMLlYBZgjKwXvGPOcUeD5xQF/UAX79/A8OUaE6djE2kPI5xPrHk7DH/a5IDZX4xIb4GOnDnobEOW45kk/+UO0efVg2pgUUihs6QAqqug1X2PgBWAVPb0ydp2MP5dYfVgK+/bvpaNWmNvxX6xTqJwfj3MYH18oo4+4kLkVypkL1Q4CHqYj7GDn8sblfu14nFaxJbvReJmPcs1ZXLfgZreBMnBE95SgQp+LPPc5SC9y/epXgkcqHI8964/S5rrM43nMt9b0hpTFurN2+z2+zdBx59oMt4hkGntbdPAT0rSsFNnOE5BwdujretRY/+RbGr2n8Er3L/iUZNnsMoxj86SS/c)

2) If we want to kill all the DOM nodes, we need an `v-if` too! However, that breaks transitions as the DOM is patched: [Vue Playground](https://play.vuejs.org/#eNqFUsmO00AQ/ZWiL5NIXpIBBDKeURBcEZc5cOiLl0qmiXtRd9lDiPzvVNuZTAgS44PVtb336qmO4rNz2dCjKEQZGq8cQUDq3b00SjvrCY7gcQsjbL3VcMOtN9JI01gTCLRtsYO72LEg3+NSmjKfYRiAA0LtuoqQI4Cy7omsgU3TqWZ/J8Xz+JvpIcXUBfBgd7sOYVBB1apTdMjgS+89GirgeDyRjuMEmc+YM/6Dr0xQpJjDVBqZIXSqRSmgcg4rf8IvWzXAkIZH+/Qs4szNVcc1tX2pcLi1nuOFSsAsQRlYr/jjSrHHAxfUxTjA1+/fwPBo1KpOOifgPLo6v1jBLDl/0cyJMr8wjMNABzYiNNZhy5lsWidlI9CnVUNqYJIIQ2eQAqqug1X2PgBWAVPb0ydpWMN5usNqwNenm75WTVrjb4V+sU5iMv4+JrBeXiHOeuJ9JFcsZC8Y2ERdzM+44Y/F7cr9OmHx1cRV70Ui5rNLdeWyn8EaPswJgl2eCkEKvoLZRyn4HGMsxSORC0We98btd1ljdb7hWu57Q0pj2lq9eZvdZu8+sPWBLvMZBp3W3j4F9MwoBS9xBs85OfByfHwtevT/Jbvq/YvwqvYPaeRkH0Yx/gHDZDS4)

3) This is not a problem of the Vue's scheduler patching the DOM before the Transition component takes the change, since this is also reproducible if we add another property that tracks the animation's state: [Vue Playground](https://play.vuejs.org/#eNqFU02P0zAQ/StDLptKTbJdQKDQXRXBFXFZIQ65uMmk661jW/5IKVX/O+M4Cd2CoIfInjfznud15pR81DrvPSZlsra14dqBRef1QyV5p5VxcAKD7RJq1WnvsIEztEZ1cENFN5WsZK2kpSK+k0zAfUhOnfG4mBCDskHzN6RTDYaSiTo9VRJghy5dwHAEqnHeyJE975nwGIDzMnzpoanEwzcm5vzLRGKOaIR4C1N2DFQuPm3ODo8b2Ct5ppx1EQ0hK+jisNOCOaQbwHrrnVMSNrXg9f6+SqZWXg2HKhmyAB7VbicQem75lgvujjl88oZkXQmn02jAmfSIsoickf/RMGm546QhWYekYAVvsEqAaY3MwEYg60N8trdlwlLCqLxueA99Zp/UYXrejBGqCePtXE68fdYqQ4GUL0EugEtY3dKPkHKPRwL4RT3A569fQBJraIOPLQzMRRideKInxG6K3+1QYF1ceElX647kka2VxoYi+dBpRh6hyVjteE8igcbNJCUwIeA2f2sBmcVMefeB/rPANlYP7vy/uvZbXmdb/MnRpKtlCIbP+yWsFleM8T1h9Gn2Xqg4daFAJnZlPIYOv6d3t/rHyEUDFVp9SJZJ3K2sYzp/tkrS9g0U5PIA2CqhARnHNKFNC/cqeXJO27IovNT7XU5rU2wIK4yXjneYNarbvM7v8jfvyHrrLuM52i7bGnWwNO/PRD8s0EheULCn5uIshGn4h9hV7gvBK+wP0Wm5kvMvSAKCZg==)

4) 3 is also reproducible if we set an interval that takes less than the transition itself: [Vue Playground](https://play.vuejs.org/#eNqFU9tu2zAM/RXOL3UAX3pZsSFLiwzb6zBgKIY9+EWx6VStLQm6OMuC/PsoyfbSbFj9YEg85CF5RB6Sj0oVg8NkmaxMrbmyYNA6dV8J3iupLRxAY5tBLXvlLDZwhFbLHi4o6KISlailMBTEt4J1cOedU6sdLiZEo2hQ/wvpZYM+ZKJOD5UA2KJNFxCOQDHWaTGyFwPrHHrgmPk/FZoK3H1n3ex/6kjMEY0Qb2HyjobKxtJmb19cYK/EkXwq0TpRWy4pP9pvwfeB9yjdXCDZJwuZ7u7HZmfGlnUGM7i5JTpiXZVRYhKXLhZ71TGLdANYbZy1lGldd7x+vquSSZw34VAlwQvgQW63HcLADd/wjtt9AZ+cprR2CYfDKOmRchFlGTkj/4NmwvDQjWA9UgbT8QarBJhSyDSsO2RDsJ81O+deNXyAITePcjcVOGOEKsJ4S0gUgZiHvJWaDCnPQCyAC7i6pI+Q5TPuCeAn8QCfv34BQay+ET42EZhLP47xRCXEfso/DZFhVZ6oSVdj96SSqaXChixF6DUnlVDnjN50oCSexs4kS2BdB5fFrQFkBnNq/EN4tDk66PN6dO02vM43+IujTq8yb/S/9xlcLc4YYz1+nWieX2Sx8iQDidgv49F3+CO9vlQ/Ry4aKd/qfZIlcV/znqniyUhBGx0oSOUAmCqhERlHP6Ht9fcqebRWmWVZOqGetwWtYrkmrNROWHr+vJH9+qa4Lt6+I+mNPbUXaPp8o+XO0MQ/EX1YypG8JONAzcVZ8NPwn2Rnvi8SnmF/JZ0WNjn+Bst0oHI=)

5) The current solution everybody is doing to solve this problem is very similar to what has been done in 3) and 4): have 2 variables, one that tracks the **model** state and another that tracks the **rendering** itself. With the `after-leave` hook, we have an effective workaround: [Vue Playground](https://play.vuejs.org/#eNqFU8GO0zAQ/ZUhl02lJtkuIFDororgirisEIdc3GTS9TaxLXuSUqr+O+M4Cd2CoIfInjfznud15hR9NCbtO4zyaO1KKw2BQ+rMQ6Fka7QlOIHFegmlbk1HWMEZaqtbuOGim0IVqtTKcZHcKdHAvU+OyXa4mBCLqkL7N6TVFfqSiTo+FQpghxQvYDgC11Bn1cie9qLp0APnpf/yQ2OFh2+imfMvE5k5oAGSNUzZIVBQeNqc7R83sBfqzDnrLBjCVvCFsDWNIOQbwHrbEWkFm7KR5f6+iKZWXg2HIhqyAB71btcg9NLJrWwkHVP41FmWpRxOp9GAM+sxZRY4A/+jFcpJkqyhRIus4BpZYRGBMAaFhY2oCW3SoOg9Optci8Zx2qi/rmQPfeKe9GF65IwxahiT9VzO7H1Sa8uBWC5BLUAqWN3yj5F8j0cG5EU9wOevX0Axq29Gjo0MzJkfoHDiJ4Sest9NcWCdXTjKV0dHdsqV2mDFkXToN2GnuEtRkuxZxNPQTJKDaBq4Td86QOEw0R194H/Os43Vgzv/ry67rSyTLf6UaOPV0gf95/0SVosrxvAevwA8gS9USF8osIltHo6+w+/x3a35MXLxWPlWH6JlFDYsaYVJn51WvIMDBbs8AK6IeEzGYY143/y9iJ6IjMuzrFNmv0t5ebINY5ntFMkWk0q3m9fpXfrmHVvv6DKeomuTrdUHx1P/zPTDGo3kGQd7bi7Mgp+Gf4hd5b4QvML+EJ1WLDr/AgzGhKU=)

# Proposed solution

There are two things that could be done into the `leave` and `after-leave` hooks of `Transition` to solve this:

### Stop dependency/effect tracking of the child component recursively.
This one probably introduces a lot of edge cases and will be really difficult to implement, besides being more prone to breaking watchers, computed properties, etc... And probably will make the implementation much more difficult, since the `Transition` component itself relies in the the state of the component to know when it needs to fire the transition*

### Stop [Vue's patch process](https://vuejs.org/guide/extras/rendering-mechanism.html#render-pipeline).

Inside Vue, only `Transition` should be manipulating the DOM nodes in the "paused" state of the patcher to add the CSS styles and event handler needed for the correct functionality of the component:

· **Entering component**: 
- `sync` flush timing:
   * Patch the slot's content with the new state (the same that triggered `Transition`)
   * Apply (if necessary) the pre-transition styles
   * [...]
   * **All the enqueued DOM updates are finished and a new cycle starts**
- Stop patcher and apply event handlers and the rest of the styles that finally start the CSS transition
- *Transition starts*
- `after-leave`:
   * Resume patcher.
   * Add to the patcher's scheduler any state change during the transition.
   * Patcher applies the changes
   * [...]
   * **All the enqueued DOM updates are finished and a new cycle starts**

· **Leaving component**: 
- `pre` flush timing:
   * Stop patcher
   * Apply (if necessary) the pre-transition styles
- Apply event handlers and the rest of the styles that finally start the CSS transition
- *Transition starts*
- `after-leave`:
   * Resume patcher.
   * Add to the patcher's scheduler any state change during the transition.
   * Patcher applies the changes
   * [...]
   * **All the enqueued DOM updates are finished and a new cycle starts**

# Adoption strategy

This new behaviour should be the default for `Transition`, since I think it's what most people expect. A `patchable` | `modifiable` boolean prop (with proper documentation) could be added to the component to keep the current behaviour.

# Drawbacks

None that I'm aware of

# Alternatives

Approach 5) in the **Motivation/Summary** section.

# Additional context

* I'm not going too deep into my proposal since I just did a surface inspection of `Teleport` source code and, although I consider myself proficient enough in "userland" Vue, I'm just starting to become more and more interested in Vue's internals and there are probably a lot of things that I'm missing. **Please, forgive any lack of precision. Any detailed explanation or further precision of the fine details to get this proposal to a good point is more than welcome!**
     * That's why I used [flush timing](https://vuejs.org/guide/essentials/watchers.html#callback-flush-timing) to roughly describe how my proposition should work, as I assume the patcher and watcher's flush timing are essentially the same. 
     * Also, why I'm not aware of any drawbacks.
* I can look into implementing this myself as soon as this thread gets some feedback, but I don't have anything against somebody else taking the challenge either and doing it before me.

**Comment by ferferga** [comment]
Work that is going to help make this RFC possible is already live at https://github.com/vuejs/core/pull/9206#issuecomment-1881896826 and  https://github.com/vuejs/core/pull/9651

**Comment by ferferga** [comment]
Digging more into this, it looks like this is a common source of trouble that framework authors have been dealing, since besides the Vuetify threads linked in the main post, Radix Vue also had the same issue. They workaround this using their [``Presence`` component](https://github.com/radix-vue/radix-vue/blob/main/packages/radix-vue/src/Presence/Presence.ts)

@zernonia Do you think all the use cases of Radix will be covered by this proposal? Any edge case you think that might be missed by pausing effects while transitions are running? Given you develop a component library and have some experience with Presence, any API suggestion to have this feature upstream? Thank you very much in advance,

**Comment by jods4** [comment]
IMHO This has a big potential for edge cases and unforeseen consequences.

A few random comments, but I'm sure more will be uncovered if people experiment with this:
- While I agree that the majority of the time it's ok/expected to freeze leaving components, I am really not sure this applies to entering components as well (as your proposal suggests to also freeze entering components). I've seen several demos of the relatively new CSS view transitions where emphasis was put on "content is still live during the transition!!".
- You mention CSS transitions, but what about JS transitions (e.g. orchestrated ones)? Callbacks `onLeave` and `onBeforeLeave` are meant to be used for doing JS-driven animations and although they are rather vanilla-JS oriented (they pass an HTMLElement), nothing is preventing the user to capture Vue state or find it back from the element and attempt to update the DOM through state changes. A flag to control this behaviour is definitely needed and if we want to avoid potentially breaking changes it should remain "live" by default.
- How does this interact with interruptible animations? E.g. consider the dropdown list of a combobox, could be driven by a `v-if` or maybe `v-show`. If the user clicks quickly you want to interrupt the ongoing animation and "invert" the animation from current state (ideally preserving velocity, MotionOne feels so good for those things). If you're not patching the state during the leave animation, is this still going to work?
- Thinking long-term: leave transitions are inherently hard and hacky; CSS view transitions are going to be a much better solution for this. Right now they only support singleton (document-level) transitions but there's ongoing work to extend them to element-level transitions. After this lands, this kind of complicated hacks will be unnecessary, so is it worth investing in those or can we continue as-is for a few more years?

**Reply by ferferga** [reply]
@jods4 Apologize for taking so long to reply, been busy these weeks and I've been delaying this more and more.

> While I agree that the majority of the time it's ok/expected to freeze leaving components, I am really not sure this applies to entering components as well (as your proposal suggests to also freeze entering components). I've seen several demos of the relatively new CSS view transitions where emphasis was put on "content is still live during the transition!!".

This doesn't change at all with the current behaviour (unless I'm misunderstanding something). When a component is enabled by ``v-if``, the entire VNode tree of that component is appended to the DOM and then transition starts. If it didn't happen in that order, what DOM node would the ``Transition`` component append the CSS classes/JS events to?
With my proposal it would work like this:

* Vnode is mounted with the equivalent behaviour of the ``v-once`` directive
* Transition plays
* ``v-once`` is turned off

> You mention CSS transitions, but what about JS transitions (e.g. orchestrated ones)? Callbacks onLeave and onBeforeLeave are meant to be used for doing JS-driven animations and although they are rather vanilla-JS oriented (they pass an HTMLElement), nothing is preventing the user to capture Vue state or find it back from the element and attempt to update the DOM through state changes. A flag to control this behaviour is definitely needed and if we want to avoid potentially breaking changes it should remain "live" by default.

I don't understand what's the issue with this, can you give an example? If an user wants to manipulate the DOM outside Vue, of course nothing we can do to stop them from doing so, but that's irrelevant!
The current behaviour that happens when you manually modify the DOM outside Vue will happen in this case: once the patcher needs to patch the element, Vue's VDom will overwrite those changes as it will unaware that they happened in the first place.

> How does this interact with interruptible animations? E.g. consider the dropdown list of a combobox, could be driven by a v-if or maybe v-show. If the user clicks quickly you want to interrupt the ongoing animation and "invert" the animation from current state (ideally preserving velocity, MotionOne feels so good for those things). If you're not patching the state during the leave animation, is this still going to work?

Is as simple as: Transition happening -> patcher paused. Transition not happening -> patcher running

So in your example, a quick ``v-if`` sucession of ``false - true -false`` would:

* Mount component as v-once.
* Transition appends enter animation classes and events, pauses patcher
* Component is toggled false
* Enter transition finished, patcher resumes
* Transition appends close animation classes and events, pauses patcher
* Transition ends, component is unmounted

This is the same behaviour as [now](https://play.vuejs.org/#__PROD__eNp9Uk1P3DAQ/SuDL7QSm0Apl21A/RBS20NbAVIvlirXO7sxOLZlT9JFq/z3juNVWCTEzfNm3vPzG+/EpxCqoUexFE3S0QSChNSHK+lMF3wk2EFqlbX+3w2uYYR19B0cM+NYOum0d4kZZuOUhcuDyTcUe3z7QbqmLrIsyAVhF6wi5Aqg+dsTeQcftTX64VKKWeeonKS4up0OS9ixjdIdx6YuxCJyF5VLhsy+ZqQ9g2Fh1rMgy3xF9gW/fbSro6Zuzwq1fsZt6gN3XCZ6tAhJ+4ArRqphgY4wLpQmM+DJhFhUA+4R2GVRmiWX4IPShh7hIgGqhJzGmJVnpZzloQ75orHnLeG0UDjD7IVNiBNB7Mitzaa6T97x2iaGFNp3wViMP0O+O0nBmZU8pJiW8n3C8lr4yoLrFvXDC/h92mZMil8RE8YBpZh7pOIGqbSvb3/gls9zs/Or3vL0K80bTN722WMZ+9y7Fds+mJvcfps+n3Gbu3S9JeRI94/KRvPkOM1LwT/xyytPf7J7Xr2feJwop/hnwJg1OcDz6qI6XSgbWlW9E+N/v2QKnw==)

Either case, this being opt-in means that users can simply skip adding the ``freeze`` prop if they don't like the behaviour.

> Thinking long-term: leave transitions are inherently hard and hacky; CSS view transitions are going to be a much better solution for this. Right now they only support singleton (document-level) transitions but there's ongoing work to extend them to element-level transitions. After this lands, this kind of complicated hacks will be unnecessary, so is it worth investing in those or can we continue as-is for a few more years?

This is a really good point, though I didn't have enough time to play with the API yet to see how much this would help in all situations. I understand your point about the main point of CSS View Transition API is to have content live while it's transitioning, but this is not what is always desired. My Vuetify playground is the best example of this, I don't want it to go white at all, just fade away! Yes, I could change the color on an @afterLeave, but that's way too verbose.

Beyond the usability, I think another interesting point that we should bring into the table is the performance improvements it could bring: pausing activity and manipulation of the DOM in the main thread while a transition is working and it's undesired will always be a welcome thing that View Transition API will never solve because it's completely out of it's scope.
